org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/XMLTextDocumentService.java

```diff
@@ -206,29 +209,38 @@ private XMLFormattingOptions getFormattingSettings(String uri) {
 			DocumentSymbolParams params) {
 
 		TextDocument document = getDocument(params.getTextDocument().getUri());
+		XMLSymbolSettings symbolSettings = sharedSettings.getSymbolSettings();
 
-		if (!sharedSettings.getSymbolSettings().isEnabled()
-				|| sharedSettings.getSymbolSettings().isExcluded(document.getUri())) {
+		if (!symbolSettings.isEnabled()
+				|| symbolSettings.isExcluded(document.getUri())) {
 			return CompletableFuture.completedFuture(Collections.emptyList());
 		}
 
 		return computeDOMAsync(params.getTextDocument(), (cancelChecker, xmlDocument) -> {
-			if (hierarchicalDocumentSymbolSupport) {
-				return getXMLLanguageService().findDocumentSymbols(xmlDocument, cancelChecker) //
-						.stream() //
-						.map(s -> {
-							Either<SymbolInformation, DocumentSymbol> e = Either.forRight(s);
-							return e;
-						}) //
-						.collect(Collectors.toList());
-			}
-			return getXMLLanguageService().findSymbolInformations(xmlDocument, cancelChecker) //
+			// if (hierarchicalDocumentSymbolSupport) {
+			// 	return getXMLLanguageService().findDocumentSymbols(xmlDocument, symbolSettings, cancelChecker) //
+			// 			.stream() //
+			// 			.map(s -> {
+			// 				Either<SymbolInformation, DocumentSymbol> e = Either.forRight(s);
+			// 				return e;
+			// 			}) //
+			// 			.collect(Collectors.toList());
+			// }
+			List<Either<SymbolInformation, DocumentSymbol>> symbolInfo =  getXMLLanguageService().findSymbolInformations(xmlDocument, symbolSettings, cancelChecker) //
 					.stream() //
 					.map(s -> {
 						Either<SymbolInformation, DocumentSymbol> e = Either.forLeft(s);
 						return e;
 					}) //
 					.collect(Collectors.toList());
+			int maxItems = symbolSettings.getMaxItemsComputed();
+			if (symbolInfo.size() > maxItems) {
+				String message = xmlDocument.getLocalName() +
+						": For performance reasons, document symbols have been limited to " + maxItems + " items. ";
+				xmlLanguageServer.getLanguageClient().showMessage(new MessageParams(MessageType.Warning, message));
+				return symbolInfo.stream().limit(maxItems).collect(Collectors.toList());

```

Comment:
Maybe its not very good to do this.

The way its implemented right now, is that if the `maxItemsComputed` is exceeded, we would know by finding the length of the resulting list. The resulting list would only exceed `maxItemsComputed` by at most 1.

Reply:
I don't think this strategy will work very well for `documentSymbols` because the document symbols provides a hierarchy.

In other words, `getXMLLanguageService().findDocumentSymbols(xmlDocument, cancelChecker).size()` is not the number of symbols.

Reply:
@xorye see my draft PR which should fix your 2 issues